### ISSUE TITLE
[$250] User validation –User logging out when validate account in NewDot that was created in OldDot

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#41330]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.4.68-0
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**If this was caught during regression testing, add the test name, ID and link from TestRail:** https://expensify.testrail.io/index.php?/tests/view/4524500
**Email or phone of affected tester (no customers):** ponikarchuks+830424@gmail.com
**Issue reported by:** Applause - Internal Team

## Action Performed:
1. Log out of both NewDot and OldDot
2. Navigate to OldDot https://staging.expensify.com/
1. In the login screen enter a new gmail account
2. Verify you're navigated to staging NewDot
3. Navigate to account Settings > Contact methods
4. Click on the email
5. Verify you're prompted for the magic code to validate the account
6. Verify the account receives a magic code to their email inbox
7. Enter the magic code in the field
8. Click on Verify

## Expected Result:
The account is validated

## Actual Result:
User logging out when try to validate account in NewDot that was created in OldDot. On step 6 User receive a magic code only after resend it

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

</details> 

https://github.com/Expensify/App/assets/78819774/36a14f81-5c12-4c25-b18e-72f77875a4cd

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~0126755b0189ef4aa2</li>
        <li>Upwork Job ID: 1785612522728562689</li>
        <li>Last Price Increase: 2024-05-29</li>
<li>Automatic offers: </li>
<ul>
<li>c3024 | Reviewer | 102590991</li>
</ul></li>
    </ul>
</details>

**Comment by lanitochka17** [comment]
@sonialiap  FYI I haven't added the External label as I wasn't 100% sure about this issue. Please take a look and add the label if you agree it's a bug and can be handled by external contributors

**Comment by humanitiesclinic** [comment]
I tried this, and did not receive any verification code with humanitiesclinic+exp@gmail.com

**Comment by humanitiesclinic** [comment]
Contributor details
Your Expensify account email: humanitiesclinic@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~01dd0a657e935bad4f?viewMode=1

**Comment by sonialiap** [comment]
Looking for proposals

**Comment by sonialiap** [comment]
looking for proposals, will retest later this week

**Comment by c3024** [comment]
Waiting for proposals.

**Comment by c3024** [comment]
No proposals yet.

**Comment by narefyev91** [comment]
Hey! I'm Nicolay from Callstack - will investigate current issue - please assign it for me. Thanks!

**Comment by c3024** [comment]
PR in draft.

**Comment by narefyev91** [comment]
@c3024 yeah - i just add draft PR to fix one part of the issue (logout needs to be investigating a bit more) - currently moved to urgent wave project - will be back to this one after(seems like it's not really urgent) finishing some tasks from wave.

**Comment by c3024** [comment]
[Not overdue](https://github.com/Expensify/App/issues/41330#issuecomment-2136665026)

**Comment by sonialiap** [comment]
Not overdue, assigned Nicolay :)

**Comment by sonialiap** [comment]
@narefyev91 just checking in if you've had a chance to get back to this one? :)

**Comment by narefyev91** [comment]
> @narefyev91 just checking in if you've had a chance to get back to this one? :)

Unfortunately not yet - still working on wave for payments - will keep you posted - when will get some free time to solve this one :-)

**Comment by teneeto** [comment]
Hey! I'm Eto Olei from Callstack. I will continue with this issue. Please assign it to me. Thanks!

**Comment by sonialiap** [comment]
Thanks Eto, adding you :)

**Comment by sonialiap** [comment]
@narefyev91 @teneeto any movement on this issue? :)

**Comment by teneeto** [comment]
Hi Sonialiap, I'm still looking into this and will drop another update today.

**Comment by sonialiap** [comment]
Great, thanks!

**Comment by teneeto** [comment]
Hi @sonialiap, Still looking into this; my aim is to get it done early hours of Monday, although I have a planned leave for next week, but I'll spend some time to see to it else I'm sure @narefyev91  will backup.

**Comment by sonialiap** [comment]
Sounds good, thanks for working on this

**Comment by sonialiap** [comment]
@teneeto happy Monday! Any updates on this? :)

**Comment by teneeto** [comment]
Hi Sonia, I'm back from leave today. @narefyev91 didn't get a chance to continue in my absence, as he also had a lot on his hands. I'm working on this today.

**Comment by sonialiap** [comment]
Thanks!

**Comment by sonialiap** [comment]
Eto and I cannot reproduce the bug of being logged out when verifying email. We are able to successfully verify our emails without being logged out.

But we both noticed that the magic link email isn't sent on step 8. We both had to click the "didn't receive" link to prompt the email (it goes through on the re-prompt)

**Comment by teneeto** [comment]
> Eto and I cannot reproduce the bug of being logged out when verifying email. We are able to successfully verify our emails without being logged out.
> 
> But we both noticed that the magic link email isn't sent on step 8. We both had to click the "didn't receive" link to prompt the email (it goes through on the re-prompt)

Correct! and we also realised that users are redirected from old-dot to new-dot on only public domains. attached a record below.

https://github.com/user-attachments/assets/9f725093-fc55-408c-bde6-1a8944c70b0f

**Comment by teneeto** [comment]
The [Draft PR](https://github.com/Expensify/App/pull/42626) @narefyev91 opened earlier, should solve for the magic link email not sent on step 8. But for review follow-up sake, I'll open a new PR. Thanks cc @sonialiap

**Comment by teneeto** [comment]
## Proposal
### Please re-state the problem that we are trying to solve in this issue.
There are basically two issues of concern here.
1. A User is logged out when they try to validate their account in NewDot that was created in OldDot (This is not reproducible as shared [here - issuecomment-2228030041](https://github.com/Expensify/App/issues/41330#issuecomment-2228030041)
2. when you create and login a new account from newDot and try to validate account email as `Contact method` from `Profile` in `Settings`, you don't receive a validation code (a magic sign-in code) for the first time or immediately except you click on `didn't receive` to trigger an email.


### What is the root cause of that problem?
There is currently no logic or function that triggers API to request or send code when the user opens validate code form for the first time.

https://github.com/user-attachments/assets/1258cc7c-07f7-464f-96f0-faad1f05ecfd


### What changes do you think we should make in order to solve the problem?
We can add an effect with logic on `BaseValidateCodeForm` component to firstly check if magic code has been sent before or at least if the initial code has been sent or not, with this we can trigger an API to send code if the statement is false.

An effect is needed because we to watch for changes on first load, also we need a useRef which will serve as a register to mark if the initial code has been sent or not.

**_Here is a snippet for better context_** 

Add the following code to [BaseValidateCodeForm](https://github.com/Expensify/App/blob/main/src/pages/settings/Profile/Contacts/ValidateCodeForm/BaseValidateCodeForm.tsx)

- Set a ref at top of function component
```
const isInitialCodeSent = useRef<boolean>(false);
```

- include this around were other effects are used
```
    useEffect(() => {
        if (isInitialCodeSent.current || hasMagicCodeBeenSent || (!hasMagicCodeBeenSent && loginData?.pendingFields?.validateCodeSent)) {
            return;
        }
        isInitialCodeSent.current = true;
        User.requestContactMethodValidateCode(contactMethod);
    }, [contactMethod, hasMagicCodeBeenSent, loginData?.pendingFields?.validateCodeSent]);
```

<details>
<summary>Result</summary>



https://github.com/user-attachments/assets/4ea7afa3-937f-400c-8dc9-396d4a06f832


</details>


### What alternative solutions did you explore? (Optional)
NIL

**Comment by teneeto** [comment]
It is still in review, Melvin.

**Comment by c3024** [comment]
@teneeto 

This was unfortunately reverted here #46846 . Reason here #46723.

I think this seems be one of the cases of ["You might not need an Effect"](https://react.dev/learn/you-might-not-need-an-effect) we missed. 😃 

Instead of using `useEffect` I think we can attach the API request to `onPress` here 
https://github.com/Expensify/App/blob/53eee53ac8bfbdfd4d8eeaf16940e3b165483373/src/pages/settings/Profile/Contacts/ContactMethodsPage.tsx#L83
something like this
```js
                    onPress={() => {Navigation.navigate(ROUTES.SETTINGS_CONTACT_METHOD_DETAILS.getRoute(partnerUserID));
                        if (!login?.validatedDate) {
                            User.requestContactMethodValidateCode(loginName);
                        }
                    }}
                   
```

**Comment by teneeto** [comment]
> @teneeto
> 
> This was unfortunately reverted here #46846 . Reason here #46723.
> 
> I think this seems be one of the cases of ["You might not need an Effect"](https://react.dev/learn/you-might-not-need-an-effect) we missed. 😃
> 
> Instead of using `useEffect` I think we can attach the API request to `onPress` here
> 
> https://github.com/Expensify/App/blob/53eee53ac8bfbdfd4d8eeaf16940e3b165483373/src/pages/settings/Profile/Contacts/ContactMethodsPage.tsx#L83
> 
> 
> something like this
> ```js
>                     onPress={() => {Navigation.navigate(ROUTES.SETTINGS_CONTACT_METHOD_DETAILS.getRoute(partnerUserID));
>                         if (!login?.validatedDate) {
>                             User.requestContactMethodValidateCode(loginName);
>                         }
>                     }}
>                    
> ```
> 
> or have an empty dependency array for the `useEffect`.

Ok I can take a look again.

**Comment by c3024** [comment]
Tagging all involved: @marcaaron, @trjExpensify, @jasperhuangg, and @thienlnam in issue #46723 and the revert #46846 of the earlier PR #45395 for this issue.

Can you please clarify this?

When we add a contact method, the backend sends an email with a validation code even before the user visits the validation code form.

If the user doesn't verify immediately, and later clicks on this secondary contact method, they are directed to the validation code form. However, no new magic code is sent immediately; it is only sent after the user clicks on the "Didn't receive a magic code" link, as described in the Actual Result of this issue.

> On step 6, the user receives a magic code only after resending it.

What should be the expected behaviour here?
1. Send a code right after adding the contact method, regardless of whether the user visits the validation code page. After that, do not send any code when simply visiting a contact method page. A new code should only be sent when the "Didn't receive a magic code" link is clicked on the Contact Method page. This is the existing behaviour.
2. Do not send a code right after adding the contact method. Instead, send the code only when the validation code method page for a contact method is visited. With this approach, a code is sent right after the user visits the validation code form for a secondary contact method. Further code requests can be made by clicking on the "Resend validation code" link on the page.

I think option (2) is better because adding a secondary contact method only takes the user to the Contact Methods page. The user needs to click on a contact method to go to the validation form. Therefore, on this event handler, we can request the code with `requestContactMethodValidateCode`.

Thanks!

**Comment by trjExpensify** [comment]
For clarity and not conflating two different cases, let's lay them out:

**Case 1: Signed-up, hasn't validated the primary login yet.** 

1. Signs-up 
2. Gets signed-in because we allow that unvalidated session 
3. Heads to Profile > Contact method
4. Clicks on the primary contact that's unvalidated
5. A magic code link should be sent 

**Case 2: Adds a secondary login to an existing account**

1. Heads to Profile > Contact method 
2. Clicks "Add contact method"
3. Adds a secondary login 
4. An email to notify the primary login that a secondary contact method was added to their account is sent right away for visibility/security, and includes a magic code/link to verify it (below for ref). 

<img width="621" alt="image" src="https://github.com/user-attachments/assets/09118f0c-cfa0-4d6f-8413-cf2a1b525498">

So in case 2 when adding a secondary login, the bug [here](https://github.com/Expensify/App/issues/46723) was that we were sending two magic code emails in addition, as a result of the change that was reverted:

<img width="909" alt="image" src="https://github.com/user-attachments/assets/9d26532f-804e-4e1c-976d-0e28aec45bb9">

**Comment by c3024** [comment]
Thanks, @trjExpensify.

So, I think this should be the expected behaviour. Can you please check if this is correct?

**Case 1:**
1. When we click on the unvalidated primary contact method, a security code should be sent each time.

**Case 2:**
1. When we click on a secondary contact method, a security code should not be sent on this click alone, as it was already sent when the secondary contact method was added.

**Comment by trjExpensify** [comment]
Case 1, yup! 

Case 2, not quite. I think we should: 

1. Click `New contact method`
2. Enter a new login 
3. Click `Add` 
4. Fire off the secondary login email only and navigate to the magic code input page to validate it

That way, we 1) aren't sending additional magic code emails and 2) we aren't asking the user to click the unvalidated secondary login to validate it right after adding it which would cause us to fire another one off like case 1. 

Does that make sense? CC: @Expensify/design

**Comment by sonialiap** [comment]
@johncschuster back from OOO, releasing you from buddy role :D

**Comment by sonialiap** [comment]
@c3024 we have clarification from Tom for the [expected behavior](https://github.com/Expensify/App/issues/41330#issuecomment-2319063902). Is the next step to implement and push a PR?

**Comment by teneeto** [comment]
@c3024 @trjExpensify how do we plan to proceed with this? I'm waiting for your lead 😌.

**Comment by trjExpensify** [comment]
Hoping @c3024 can provide us a video of the current behaviour, so we can discuss those next steps. 

>For the benefit of the discussion and design input, can you add a video of this flow so we can see the problem it presents please? Thanks!

**Comment by c3024** [comment]
Now, with the changes in the PR #47772, the secondary contact verification page appears only for a brief period and redirects to ContactMethodsPage. 


https://github.com/user-attachments/assets/d7645eac-5c7d-4b44-950c-965d1a5594a5

I think this needs to be fixed first. @teneeto

**Comment by teneeto** [comment]
on it.

**Comment by teneeto** [comment]
Apologies for the delay, this should be the current state for secondary login.

https://github.com/user-attachments/assets/bc566339-5c08-436a-9fb4-eabbf6f89969

**Comment by c3024** [comment]
The earlier PR was reverted because of this flow after addition of secondary contact

1. Backend sends a validate code immediately for the secondary contact
2. But since we redirect to the ContactMethodsPage, we still have to click again on the secondary contact method and another code is requested from front-end. 

This flow still remains with this PR. Here is the video.

https://github.com/user-attachments/assets/3b6f7461-2428-471f-a78e-57801c054394

We need to redirect the user not to ContactMethodsPage but to the validate code page to use the code immediately received in Step 1.

**Comment by teneeto** [comment]
More straightforward to follow now, 100% get you. I will add a fix and push back. Thanks @c3024

**Comment by teneeto** [comment]
@c3024 can you confirm this is the expected flow?

https://github.com/user-attachments/assets/dfc51326-50b3-41f1-9e31-6c840537a45b

**Comment by teneeto** [comment]
This issue PR seems to have been approved by @c3024 with a few UI comments, but I'm still not clear on how we intend to proceed with it. Do you have any comments on this?

**Comment by sonialiap** [comment]
@teneeto @c3024 I see we're still working on the fix for the revert https://github.com/Expensify/App/pull/47772, is there anything I can help with to push this along?

**Comment by waterim** [comment]
Hello, Im Artem from Callstack 
As @teneeto is OOO I will take this ticket to finish on time!

**Comment by sonialiap** [comment]
Thanks Artem!

**Comment by sonialiap** [comment]
This has been open for a solid six months, asking QA for a retest to confirm if still an issue

**Comment by sonialiap** [comment]
Not reproducible - https://expensify.slack.com/archives/C9YU7BX5M/p1734120354616929?thread_ts=1734095446.002559&cid=C9YU7BX5M
